File: lib/persistence/relations/voters.rb
module Persistence
  module Relations
    class Voters < ROM::Relation[:sql]
      schema(:voters) do
        attribute :id, Types::Serial
        attribute :name, Types::Strict::String
        attribute :email, Types::Strict::String
        attribute :address, Types::Strict::String
        attribute :category, Types::Category
        attribute :party_registration, Types::Strict::String
        attribute :only_call_for_vote_reminder, Types::Strict::Bool
      end

      def by_id(id)
        where(id: id)
      end

      def by_email(email)
        where(email: email)
      end
    end
  end
end

File: web/models/voters.ex
defmodule CanvassApp.Voters do
  use CanvassApp.Web, :model

  schema "voters" do
    field :name, :string, null: false
    field :address, :string
    field :neighborhood, :string
    field :email, :string, null: false
    field :phone, :string, null: false
    field :category, :string
    field :party_registration, :string
    field :only_call_for_vote_reminder, :boolean, default: false
    field :no_fundraising, :boolean, default: false
    field :ideas, :string

    timestamps()
  end

  @required_fields [:name, :address, :email, :phone, :category, :party_registration, :neighborhood]
  @optional_fields [:phone, :only_call_for_vote_reminder, :no_fundraising, :ideas]

  @doc """
  Builds a changeset based on the `struct` and `params`.
  """
  def changeset(struct, params \\ %{}) do
    struct
    |> cast(params, @required_fields ++ @optional_fields)
    |> validate_required(@required_fields)
    |> validate_inclusion(:category, ~w(Committed Unaware Transactional))
    |> foreign_key_constraint(:voter_id)
  end
end
